### ISSUE TITLE
Is dependency pytest-asyncio required?

### ISSUE THREAD

**Issue by Reskov** [CircleUp/aresponses#82]
Hi there! Appreciate the fantastic library.

I'm curious about whether we actually need `pytest.asyncio` as a dependency. Looks like that `aiohttp.pytest_plugin` supports coroutine executions on its own.

To make packages tests run we need to just add loop as dependency and remove all `@pytest.mark.asyncio` decorators
```python
@pytest.fixture()
async def aresponses(loop) -> ResponsesMockServer:
    async with ResponsesMockServer(loop=loop) as server:
        yield server

```

**Comment by Olegt0rr** [comment]
1. `loop` fixture is deprecated. Now it works this way: https://github.com/aresponses/aresponses/blob/a8fb95842a1646b597d9e7ec29c20ae1b3398281/aresponses/main.py#L339-L343
2. You still need `pytest.asyncio` to execute `async with` statement in `async def` coroutine.

**Comment by Reskov** [comment]
@Olegt0rr  Thanks for you answer, but I am not sure about the second point, can you explain pls.

Here is a minimal example without `pytest.asyncio`

### Create a test file with an async test

test_poc.py
```python
pytest_plugins = 'aiohttp.pytest_plugin'

async def test_fail():
    assert False
```

### Install only aiohttp and pytest
```
>>> pip install pytest aiohttp
>>> pip freeze
```

```
pip freeze
aiohttp==3.9.5
aiosignal==1.3.1
async-timeout==4.0.3
attrs==23.2.0
exceptiongroup==1.2.1
frozenlist==1.4.1
idna==3.7
iniconfig==2.0.0
multidict==6.0.5
packaging==24.1
pluggy==1.5.0
pytest==8.2.2
tomli==2.0.1
yarl==1.9.4

```

### Run tests
```
>>> pytest test_poc.py
```
This will produce

```
___________________________________________________________________________ test_fail ____________________________________________________________________________

    async def test_fail():
>       assert False
E       assert False

test_poc.py:4: AssertionError
```

Am I missing something?

**Comment by Olegt0rr** [comment]
You are right, the `aiohttp.pytest_plugin` plugin really helps to run tests without connecting `pytest-asyncio`, but the latter option is a more obvious and popular way of testing asynchronous code.

Of course, let the author decide what to do, but I don't see a problem with this dependency

**Comment by Reskov** [comment]
I almost forgot my concrete case since the month passed, it was something with a event loop juggling because `aiohttp` and `pytest-asyncio` how their own. I will try to prepare a detailed example later. 

Anyway, adding these lines https://github.com/aresponses/aresponses?tab=readme-ov-file#working-with-pytest-aiohttp to each project is not obvious. And the reason for that is only to silence the package that I do not need. That doesn't look good for me as well as having unnecessary depency. What will be better is to make pytest-asyncio as optional dependency, something like `aresponse[pytest-asyncio]`

**Comment by Olegt0rr** [comment]
Docs for this library should be updated
You don't need do [this](https://github.com/aresponses/aresponses?tab=readme-ov-file#working-with-pytest-aiohttp) since v3

**Comment by Reskov** [comment]
@Olegt0rr I do, I am not able to make it work without this, because `aresponses` fixture depends on `pytest-asyncio` as I said above with a different event loop.

I suppose this happened because of 
https://github.com/aresponses/aresponses/blob/master/aresponses/main.py#L8-L10

**Comment by brycedrennan** [comment]
Apologies, I haven't had to deep dive into this library for years, so I may be a little rusty.

Having re-read the docs just now I believe aresponses is intended to work with more than just aiohttp.  I also don't like that you have to add that little snippet to work with aiohttp but I couldn't find a way to make it "just work".   That was ~6 years ago though so very possible there is a better way now.

**Comment by Reskov** [comment]
Please give it a try. Am I missing something?
```python
import asyncio

import aiohttp
import pytest

pytest_plugins = 'aiohttp.pytest_plugin'

@pytest.fixture(scope="session")
def loop():
    # use single loop on all session
    return asyncio.get_event_loop()


@pytest.fixture(scope="session")
def event_loop(loop):
    # for aresponses
    return loop


async def test_ok(aresponses):
    aresponses.add('foo.com', '/', 'get', 'hi there!!')
    async with aiohttp.ClientSession() as session:
        async with session.get('http://foo.com') as response:
            text = await response.text()
            assert text == 'hi'
```


### This will stack forever

```python
import asyncio

import aiohttp
import pytest

pytest_plugins = 'aiohttp.pytest_plugin'

async def test_ok(aresponses):
    aresponses.add('foo.com', '/', 'get', 'hi there!!')
    async with aiohttp.ClientSession() as session:
        async with session.get('http://foo.com') as response:
            text = await response.text()
            assert text == 'hi'
```

**Comment by Reskov** [comment]
@brycedrennan Can we have `pytest-asyncio` as an optional dependency?  And slightly different aresponses fixtures,
- with a different decorator like we already have https://github.com/aresponses/aresponses/blob/master/aresponses/main.py#L8-L10
-  looks like for aiohttp test pluging fixtures should depends on `loop` so we should have a different signature.

**Comment by brycedrennan** [comment]
>  aresponses is intended to work with more than just aiohttp
Okay reading the code this can't be true since it works by mocking out parts of aiohttp

> Can we have pytest-asyncio as an optional dependency
I'm not sure. What is best practice these days for writing asyncio tests?

**Comment by Reskov** [comment]
> What is best practice these days for writing asyncio tests?

I think that I am not the right person to answer 🙂 but I would say that `pytest-asyncio` is quite popular 🤷‍♂️.  I suppose the main reason is name, easier to search. I personally prefer aiohttp test plugin, since it is embedded and no need to put `pytest.mark.asyncio` everywhere.

 The `aiohttp` and `anyio` is another alternative, but sadly for `aiohttp` https://docs.aiohttp.org/en/stable/testing.html and `anyio` https://anyio.readthedocs.io/en/stable/testing.html
 it is hard to measure popularity since test plugin is embedded inside the main package.

- https://pypistats.org/packages/pytest-aiohttp (low 40k daily)
- https://pypistats.org/packages/pytest-asyncio  (high 600k)
- https://pypistats.org/packages/anyio (unknown)
- https://pypistats.org/packages/aiohttp (unknown)

P.S. I rethought, and I would **not** call the numbers for pytest-aiohttp **low**, since it's only for `aiohttp` and `pytest-asyncio` for any kind of `async` code.

**Comment by brycedrennan** [comment]
Well after looking through the docs and [open issues](https://github.com/pytest-dev/pytest-asyncio/issues/868) for pytest-asyncio and pytest-aiohttp...

Dang. crazy this stuff isn't sorted out six years later :-) I do not miss it. 

I don't love that pytest-asyncio appears to have [significant bugs](https://github.com/pytest-dev/pytest-asyncio/issues/868) and I don't like that aiohttp rolls its own asyncio test framework.  

If I recall correctly, I made this design choice because I needed to test other async things outside of aiohttp in my same test suite and the aiohttp testing framework had no options because its only for testing aiohttp.  (again vague memory) but I was peeved that the aiohttp assumed the only thing i'd be testing is it's own stack.

**Comment by brycedrennan** [comment]
So I think the idea was that the default setting for aresponses is to not force the user into using the overly-needy aiohttp testing framework.

**Comment by Reskov** [comment]
>So I think the idea was that the default setting for aresponses is to not force the user into using the overly-needy aiohttp testing framework.

But now the library forced to use `pytest-asyncio` :) For me as a server developer I need aiohttp testing framework everywhere since I wanted to start my app anyway.

I agree it's odd to use `aiohttp` tests for things outside `aiohttp` and to rely on `pytest-asyncio` when it's not necessary. It's better to support both, so users can choose what's best for their needs. This way, the test suite stays flexible.

**Comment by brycedrennan** [comment]
again just fuzzy memories-- but the issue was non-aiohttp asyncio tests were not even possible while using the aiohttp library.  using pytest-asyncio doesn't break aiohttp tests but the reverse was not true... (maybe 😄 )

**Comment by brycedrennan** [comment]
or maybe it was that I would have had to re-write my entire test suite of non-aiohttp-related tests because they have their own way of doing things. something like that, either way I hated the presumption that their one library should dictate totally unrelated tests
